[Quant] Add dynamic quantization config for x86 inductor backend

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115337

**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_qat_dynamic_quant_linear
```


cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
Add quantizer for x86 inductor dynamic quant

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -216,6 +218,42 @@ def get_default_x86_inductor_quantization_config(is_qat: bool = False):
+        quant_min=0,
+        quant_max=255,
+        is_dynamic=True,
+        observer_or_fake_quant_ctr=PlaceholderObserver.with_args(is_dynamic=True),
```
this is PTQ only? maybe you can add this to the name of this util

Xia-Weiwen replied:
Thanks. I have added QAT support.

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -216,6 +218,42 @@ def get_default_x86_inductor_quantization_config(is_qat: bool = False):
+        is_dynamic=False,  # Weight is not quantized at runtime but ahead of time
+        observer_or_fake_quant_ctr=PerChannelMinMaxObserver.with_args(**extra_args),
+    )
+    bias_observer_or_fake_quant_ctr: _ObserverOrFakeQuantizeConstructor = (
```
this can be None

Xia-Weiwen replied:
Thanks. I have changed it to None to align with XNNPACK.

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -216,6 +218,42 @@ def get_default_x86_inductor_quantization_config(is_qat: bool = False):
+    bias_observer_or_fake_quant_ctr: _ObserverOrFakeQuantizeConstructor = (
+        PlaceholderObserver
+    )
+    bias_quantization_spec = QuantizationSpec(
```
this can be None, see https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/xnnpack_quantizer.py#L166

Xia-Weiwen replied:
Thanks. I have changed it to None.

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -88,12 +88,11 @@ def _port_metadata_for_input_quant_nodes(
         _add_metadata(dynamic_q_node, node)
-        _add_metadata(dynamic_dq_node, node)
+        # q_node may have multiple users, e.g., in self-attention
+        for dynamic_dq_node in dynamic_q_node_users:
```
this is a bit weird, why does q node have multiple users for dynamic quantization? doesn't this break the `choose_qparams -> q -> dq` pattern?

Xia-Weiwen replied:
Hi @jerryzh168. In the QKV case, multiple `qd`'s are added by the `DuplicateDQPass` here: https://github.com/pytorch/pytorch/blob/9dda4b20a0dcc4786a130c12fc3e4a6a1ecf23d8/torch/ao/quantization/quantize_pt2e.py#L235C1-L238C1
I think this affects both static quant and dynamic quant. Do you have any suggestions on how to deal with this? Thanks

jerryzh168 replied:
oh I see, I think in this case we'd need to duplicate not just dq, but also the entire chain of "choose_qparams -> q -> dq". this sounds like a hacky solution though, maybe we want to redesign this. cc @kimishpatel

jerryzh168 replied:
I think it's also related to how we match the patterns, if we manually match the pattern this could work as well. see https://github.com/pytorch/pytorch/pull/83459

Xia-Weiwen replied:
Yes. For X86Inductor, we plan to match the `dq -> linear` pattern only. A single `choose_qparam -> q` with duplicate `dq`'s is what we want. If a backend wants to duplicate the entire chain of `choose_qparams -> q -> dq`, can this be done during lowering?

kimishpatel replied:
Sorry, am not reading a whole lot so just asking my question. Is the issue that dq is duplicated even for dynamic quant case?

kimishpatel replied:
>  single choose_qparam -> q with duplicate dq's is what we want

Why not the whole chain? That sounds a bit more reasonable to me.

jerryzh168 replied:
yeah I agree with @kimishpatel, I think matching the whole chain will be more robust, otherwise how do you know it's a pattern for dynamic quantization

Xia-Weiwen replied:
For X86Inductor backend, we do not need to know it's dynamic quantization actually. We only want to fuse `dq - linear` and leave `choose_qparam - q` in the graph. I have some explanation here: https://github.com/pytorch/pytorch/pull/115337/files#r1440093632
I will update this PR so that `dq` is not duplicated for dynamic quantization.

Xia-Weiwen replied:
Please review again. Thanks!

kimishpatel replied:
@Xia-Weiwen btw this may become relevant for some of the works we are doing elsewhere. Specifdicaly in LLM kind of models dynamically quantizing q, k, v projection makes sense. which is what you are trying to do. In this case for xnnpack quantizer, we will have a similar issue. My earlier suggestion to not duplicate dq for dynamic quant was based on assumtion that we want to treat "dynamically quantized linear" differently from "quantized linear". But this discussion has made me think that you can argue that "dynamically quantized linear" = "dynamic quantized act" + "static quantized linear". If we follow this, then your changes of duplicating dq for dynamic quant still makes sense. I want to think a bit more on this. @jerryzh168 thoughts?

Xia-Weiwen replied:
Thanks for the comment. I will land this PR then.

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_qat_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -88,12 +88,11 @@ def _port_metadata_for_input_quant_nodes(
         scale_node = choose_qparam_users.pop()
         dynamic_q_node = next(iter(scale_node.users.keys()))
         dynamic_q_node_users = _filter_sym_size_users(dynamic_q_node)
-        if len(dynamic_q_node_users) > 1:
```
I think maybe we just don't duplicate in this case and have the lowering pass to manually match the chain? something we did in https://github.com/pytorch/pytorch/pull/83459

Xia-Weiwen replied:
Hi @jerryzh168 Are you suggesting we modify the `convert_pt2e` path and do not duplicate `dq` in the dynamic quant case? Thanks

jerryzh168 replied:
yeah, I feel it's better to either not duplicate or duplicate "choose_qparams -> q -> dq". and not duplicate anything feels cleaner

kimishpatel replied:
Agree with @jerryzh168. My preference would be to duplicate the whole chain. I would have that when i added duplicate dq, it was already doing that but maybe not.

My rationale for duplicating whole chain is that the definition of say "DQ linear" is "fp32-> chose_qparams -> q -> dq ....". So if you ever pattern match to find DQ linear, you want to match that. But this argument is solely driven from pattern matching perspective.

Xia-Weiwen replied:
Hi @jerryzh168 @kimishpatel May I know whether you have reached an agreement? Jerry was suggesting not duplicating `choose_qparam -> q -> dq` while Kimish's preference is to duplicate the whole chain. What is the conclusion?

kimishpatel replied:
@Xia-Weiwen can you say more on why do you have a pattern where dynamic quant node iis consumed by multiple nodes? Are you dynamically quantizing output of a node? And then there are multiple consumers? In this case duplicating the entire chain can result in sub optimal performance in inductor backend? While duplicating only dq still allows you to share "chose qparam -> q"? If so I think what Jerry said would make sense. That is just not duplicate dq at all.

Xia-Weiwen replied:
Hi @kimishpatel. A typical case is the Q/K/V projections in self-attention. The pattern is like:
```
 linear (Q)
  /
x --- linear (K)
  \
 linear (V)
```
With current implementation of quantization, the pattern becomes the following after the `DuplicateDQPass`:
```
                     dq - linear (Q)
                     /
x - choose_qparam - q - dq - linear (K)
                     \
                     dq - linear (V)
```
This pattern with duplicate dq is what we want for the X86Inductor backend. We plan to lower it by fusing `dq - linear` only as static quantized linear:
```
                     static_quant_linear (Q)
                     /
x - choose_qparam - q - static_quant_linear (K)
                     \
                     static_quant_linear (V)
```
Because we don't want to include `choose_qparam` and `q` in the kernel, and we can reuse the code for static quantization.
As you have agreed, let's not duplicate `dq` at all for dynamic quantization. The X86Inductor backend can duplicate `dq` when doing the lowering. How does that sound to you? Thanks.

Xia-Weiwen replied:
I have updated this PR. Please review again. Thanks!

kimishpatel replied:
> As you have agreed, let's not duplicate dq at all for dynamic quantization. The X86Inductor backend can duplicate dq when doing the lowering. How does that sound to you? Thanks.

Yeah that makes sense.

BTW how are you doing annotation for quantization for this pattern? Is it quantizing X, i.e. output of attention layer, via dynamic quantization OR is it doing dynamic quantization of q, k, v linear projections in attention? Former is dynamic quant on output, the latter is on input.

jerryzh168 replied:
@kimishpatel it's probably annotated with this: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/x86_inductor_quantizer.py#L932

basically the partition does not know about connection information so it will just find one linear pattern in
```
x -> linear1
 \ -> linear2
 \ -> linear3
```
and annotate it

Xia-Weiwen replied:
Hi @jerryzh168 Why did you say it would just find one linear pattern? I think all three linear patterns are found and annotated.

kimishpatel replied:
@Xia-Weiwen my question is whether you are annotating input activation with dynamic quant config for all 3 linears. If so I would have expected to see 3 chains of chose qparams -> q -> dq.

Xia-Weiwen replied:
I see. We see three chains of `dq - linear`.

jerryzh168 replied:
> Hi @jerryzh168 Why did you say it would just find one linear pattern? I think all three linear patterns are found and annotated.

oh sorry, I think it will find three patterns and annotated three times as well, but we only insert one observer for input because of the deduplication logic

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_qat_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -120,10 +120,14 @@ def get_input_act_qspec(quantization_config: Optional[QuantizationConfig]):
+            torch.per_tensor_symmetric,
+        ]
+        or quantization_spec.is_dynamic
+    ), "Invalid qscheme for activation in quantization spec"
```
are you using this util? this is supposed to be xnnpack only, so I think you should copy paste this util function

Xia-Weiwen replied:
We are using it. I found this change is not needed. So, I just reverted it.

jerryzh168 reviewed (approved):
LGTM

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_qat_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add dynamic quantization config for x86 inductor backend"


**Description**
Add dynamic quantization config for x86 inductor backend.
To support the QKV structure in self-attention, we removed an assertion in port-metadata-pass that requires single dequantize node after quantize node.

**Test plan**
```
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_dynamic_quant_linear
python test/test_quantization.py -k TestQuantizePT2EX86Inductor.test_qat_dynamic_quant_linear
```


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge
